src/components/Form.js

```diff
@@ -0,0 +1,214 @@
+import React from 'react';
+import {ScrollView, View} from 'react-native';
+import PropTypes from 'prop-types';
+import _ from 'underscore';
+import {withOnyx} from 'react-native-onyx';
+import compose from '../libs/compose';
+import withLocalize, {withLocalizePropTypes} from './withLocalize';
+import * as FormActions from '../libs/actions/FormActions';
+import styles from '../styles/styles';
+import FormAlertWithSubmitButton from './FormAlertWithSubmitButton';
+
+const propTypes = {
+    /** A unique Onyx key identifying the form */
+    formID: PropTypes.string.isRequired,
+
+    /** Text to be displayed in the submit button */
+    submitButtonText: PropTypes.string.isRequired,
+
+    /** Callback to validate the form */
+    validate: PropTypes.func.isRequired,
+
+    /** Callback to submit the form */
+    onSubmit: PropTypes.func.isRequired,
+
+    children: PropTypes.node.isRequired,
+
+    /* Onyx Props */
+
+    /** Contains the form state that must be accessed outside of the component */
+    formState: PropTypes.shape({
+
+        /** Controls the loading state of the form */
+        isSubmitting: PropTypes.bool,
+
+        /** Server side error message */
+        serverErrorMessage: PropTypes.string,
+    }),
+
+    // eslint-disable-next-line react/forbid-prop-types
+    draftValues: PropTypes.object,
+
+    ...withLocalizePropTypes,
+};
+
+const defaultProps = {
+    formState: {
+        isSubmitting: false,
+        serverErrorMessage: '',
+    },
+    draftValues: {},
+};
+
+class Form extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            errors: {},
+        };
+
+        this.inputRefs = {};
+        this.touchedInputs = {};
+
+        this.getValues = this.getValues.bind(this);
+        this.setTouchedInput = this.setTouchedInput.bind(this);
+        this.validate = this.validate.bind(this);
+        this.submit = this.submit.bind(this);
+    }
+
+    /**
+     * @param {String} inputID - The inputID of the input being touched
+     */
+    setTouchedInput(inputID) {
+        this.touchedInputs[inputID] = true;
+    }
+
+    /**
+     * @returns {Object} - An object containing the values for each inputID
+     */
+    getValues() {
+        const values = {};
+        _.each(this.inputRefs, (inputRef, inputID) => {
+            values[inputID] = inputRef.value;
+        });
+        return values;
+    }
+
+    submit() {
+        // Return early if the form is already submitting to avoid duplicate submission
+        if (this.props.formState.isSubmitting) {
```

Comment:
I think there could be a race condition here:

1. User presses submit
1. We dispatch `FormActions.setIsSubmitting`
1. User presses submit again, and we find ourselves here again before all of the following happens:
    1. Onyx picks up `FormActions.setIsSubmitting` and writes that data to disk
    1. `withOnyx` notices that update and passes the updated prop to the parent component.
    1. The parent component re-renders (and re-renders this child component` with the updated `formState` prop.
1. Because all that hasn't happened yet, we re-submit the form with duplicate data.

Reply:
Very interesting. This seems possible, but can we reproduce it? Probably if we double-tap fast we can figure out whether it is a problem.

Reply:
This seems like a possibility. I ran a few tests and couldn't reproduce it, but maybe I'm just not fast enough 🤣  Maybe running a script would confirm this?

Reply:
I am of the opinion that we should ensure it's _not_ possible, even if we aren't able to reproduce it now. Like, maybe it's only reproducible on really slow devices or something. Race conditions are a PITA, and they can be really hard to diagnose if they do create problems.

Reply:
If there is some issue then should we look at that as an Onyx bug? It seems reasonable to assume that if we are setting data in Onyx it should happen relatively fast enough to prevent the behavior we are concerned about. Also, we are using a JS cache in the Onyx layer right so this update should be very fast on all platforms?

Reply:
Ooh didn't know about the JS cache in Onyx. So does that mean that storing a local `isSubmitting` variable in Form would result in the same outcome (since we also set that in Onyx before writing to disk)?

Reply:
>storing a local isSubmitting variable in Form would result in the same outcome

Ah sorry didn't quite understand the question. Can you give an example of what the alternative is and what is meant by "outcome"?

Reply:
Ah well, no they wouldn't have the _exact_ same behavior. A local instance variable would get set _immediately_ and on the next submit function call (e.g. if two calls to `submit()` happened in the same tick of the event loop). Onyx data has to go through a promise (even with the cache IIRC) so the update will be moved to the bottom of the call stack sort of like how a `setTimeout()` would work. 

Main point I wanted to make before is that all this happens at the speed of JS which I'd guess is generally much faster than someone tapping. Happy to be proven wrong on that one though - if there's a bug let's fix it!

Reply:
I was able to reproduce a race condition by simply adding multiple calls to `this.submit` to the `onSubmit` handler [here](https://github.com/Expensify/App/blob/98ab01eddbccf827f2fb616b2bb8a6d47b968241/src/components/Form.js#L203). I added a local `isSubmitting` variable and it got rid of the race condition. Thanks for spotting this @roryabraham!

Reply:
I think we can wait until this becomes a problem or solve it at the Onyx level. The code we have added to get around this seems more complex than before and sets a possible precedent that issues like this should be handled with similar workarounds.

There might be something here, but can we maybe investigate it separately before complicating things.

Reply:
Marc and I discussed 1:1, and I was only able to reproduce the race condition when calling `this.submit` on the same event loop. I wasn't able to reproduce the race condition when calling it with a 1ms delay. For this reason and Marc's comment above, I'll revert these changes and mark this as resolved.

Reply:
Okay, that's fine. I'm going to try to commit this to memory in case we do see any "form submits twice" issues pop up. Solving it at the Onyx level sounds like a good plan if it does come up. Thanks for the thorough investigation @luacmartins!